tools/topology/topology2/include/components/aec.conf

```diff
@@ -0,0 +1,160 @@
+}
+
+Class.Widget."aec" {
+	#
+	# Pipeline ID for the aec widget object
+	#
+	DefineAttribute."index" {}
+
+	#
+	# aec object instance
+	#
+	DefineAttribute."instance" {}
+
+	#include common component definition
+	<include/components/widget-common.conf>
+
+	DefineAttribute."num_audio_formats" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	#
+	# aec widget UUID
+	#
+	DefineAttribute."uuid" {
+		type	"string"
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.uuid"
+	}
+
+	DefineAttribute."cpc" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	DefineAttribute."is_pages" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	# Attribute categories
+	attributes {
+		#
+		# The aec widget name would be constructed using the index and instance attributes.
+		# For ex: "aec.1.1" or "aec.10.2" etc.
+		#
+		!constructor [
+			"index"
+			"instance"
+		]
+
+		#
+		# mandatory attributes that must be provided when the class is instantiated
+		#
+		!mandatory [
+			"no_pm"
+			"uuid"
+			"process_type"
+		]
+
+		#
+		# immutable attributes cannot be modified in the object instance
+		#
+		!immutable [
+			"uuid"
+			"type"
+		]
+
+		#
+		# deprecated attributes should not be added in the object instance
+		#
+		!deprecated [
+			"preload_count"
+		]
+
+		#
+		# aec widget objects instantiated within the same alsaconf node must have unique
+		# instance attribute
+		#
+		unique	"instance"
+	}
+
+	num_audio_formats 2
+	num_input_audio_formats 4
+	num_output_audio_formats 2
+	Object.Base.input_audio_format {
+		1 {
+			pin_index               0
+		}
+		2 {
+			pin_index               1
+		}
+		3 {
+			pin_index               0
+			in_channels             4
+			out_channels            4
+			in_ch_cfg       $CHANNEL_CONFIG_QUATRO
+			in_ch_map       $CHANNEL_MAP_3_POINT_1
+			out_ch_cfg      $CHANNEL_CONFIG_QUATRO
+			out_ch_map      $CHANNEL_MAP_3_POINT_1
+		}
+		4 {
+			pin_index               1
+			in_channels             4
+			out_channels            4
+			in_ch_cfg       $CHANNEL_CONFIG_QUATRO
+			in_ch_map       $CHANNEL_MAP_3_POINT_1
+			out_ch_cfg      $CHANNEL_CONFIG_QUATRO
+			out_ch_map      $CHANNEL_MAP_3_POINT_1
+		}
+	}
+
+	Object.Base.output_audio_format {
+		1 {
+			in_bit_depth            16
+			in_valid_bit_depth      16
+			out_bit_depth           16
+			out_valid_bit_depth     16
+		}
+		2 {
+			in_channels             4
+			in_bit_depth            16
+			in_valid_bit_depth      16
+			out_channels            4
+			out_bit_depth           16
+			out_valid_bit_depth     16
+			in_ch_cfg       $CHANNEL_CONFIG_QUATRO
+			in_ch_map       $CHANNEL_MAP_3_POINT_1
+			out_ch_cfg      $CHANNEL_CONFIG_QUATRO
+			out_ch_map      $CHANNEL_MAP_3_POINT_1
+		}
+	}
+
+	# Default attribute values for aec widget
+	# UUID: B780A0A6-269F-466F-B477-23DFA05AF758
+	uuid			$AEC_UUID
+	type			"effect"
+	no_pm 			"true"
+	core_id			0
+	cpc 			5000
```

Comment:
??? how is this defined?

Reply:
It is a common value in widget definition. As I know cSOF doesn't have a complete solution for cpc used in fw for schedule.  Take ref fw for example,  the cpc is zero in mod_cfg and is built from CPS (_Number of DSP cycles consumed by the module instance to process one second of data_)  in windows driver.  CPS should be generated from statistic in FW for different stream format. We can define a accurate cpc when cSOF has a solution in fw for cpc based schedule.

        [[module.entry]]
        name = "GAIN"
        uuid = "61BCA9A8-18D0-4A18-8E7B-2639219804B7"
        affinity_mask = "0x1"
        instance_count = "40"
        domain_types = "0"
        load_type = "0"
        module_type = "9"
        auto_start = "0"
        sched_caps = [1, 0x00008000]

        # pin = [dir, type, sample rate, size, container, channel-cfg]
        pin = [0, 0, 0xfeef, 0xf, 0xf, 0x45ff,
                        1, 0, 0xfeef, 0xf, 0xf, 0x45ff]

        # mod_cfg [PAR_0 PAR_1 PAR_2 PAR_3 IS_BYTES CPS IBS OBS MOD_FLAGS CPC OBLS]
        mod_cfg = [0, 0, 0, 0, 416, 914000, 48, 64, 0,                      0, 0,
                   1, 0, 0, 0, 416, 1321600, 192, 256, 0,                   0, 0,
                   2, 0, 0, 0, 416, 1786000, 192, 256, 0,                   0, 0,
                   3, 0, 0, 0, 416, 2333000, 48, 64, 0,                     0, 0,